plenum/server/i3pc_watchers.py

```diff
@@ -0,0 +1,25 @@
+from typing import Callable, Iterable
+
+
+class NetworkI3PCWatcher:
+    def __init__(self, cb: Callable):
+        self.nodes = set()
+        self.connected = set()
+        self.callback = cb
+
+    def connect(self, name: str):
+        self.connected.add(name)
+
+    def disconnect(self, name: str):
+        had_consensus = self._has_consensus()
```

Comment:
What if we face disconnection during initial establishing of connections:
- consider a pool of 4 nodes
- a node connected to Node1 
- a node connected to Node2 (so f+1 now)
- a node disconnected from Node 2 (f now) => restart???

Maybe we should do restart only if we go to <=f connection, AND we were connected to at least n-f nodes before?

Reply:
Yes, I missed that case, going to fix soon